File: src/com/maddyhome/idea/vim/regexp/CharacterClasses.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2019 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.regexp;

public final class CharacterClasses {
  private CharacterClasses() {
  }

  public static final String[] CLASS_NAMES = new String[]{
    "alnum:]",
    "alpha:]",
    "blank:]",
    "cntrl:]",
    "digit:]",
    "graph:]",
    "lower:]",
    "print:]",
    "punct:]",
    "space:]",
    "upper:]",
    "xdigit:]",
    "tab:]",
    "return:]",
    "backspace:]",
    "escape:]"
  };

  public static final int RI_DIGIT = 0x01;
  public static final int RI_HEX = 0x02;
  public static final int RI_OCTAL = 0x04;
  public static final int RI_WORD = 0x08;
  public static final int RI_HEAD = 0x10;
  public static final int RI_ALPHA = 0x20;
  public static final int RI_LOWER = 0x40;
  public static final int RI_UPPER = 0x80;
  public static final int RI_WHITE = 0x100;

  public static final int CLASS_ALNUM = 0;
  public static final int CLASS_ALPHA = 1;
  public static final int CLASS_BLANK = 2;
  public static final int CLASS_CNTRL = 3;
  public static final int CLASS_DIGIT = 4;
  public static final int CLASS_GRAPH = 5;
  public static final int CLASS_LOWER = 6;
  public static final int CLASS_PRINT = 7;
  public static final int CLASS_PUNCT = 8;
  public static final int CLASS_SPACE = 9;
  public static final int CLASS_UPPER = 10;
  public static final int CLASS_XDIGIT = 11;
  public static final int CLASS_TAB = 12;
  public static final int CLASS_RETURN = 13;
  public static final int CLASS_BACKSPACE = 14;
  public static final int CLASS_ESCAPE = 15;
  public static final int CLASS_NONE = 99;

  public static boolean isMask(char ch, int mask, int test) {
    boolean res = false;
    switch (mask) {
      case RI_DIGIT:
        res = Character.isDigit(ch);
        break;
      case RI_HEX:
        res = isHex(ch);
        break;
      case RI_OCTAL:
        res = isOctal(ch);
        break;
      case RI_WORD:
        res = isWord(ch);
        break;
      case RI_HEAD:
        res = isHead(ch);
        break;
      case RI_ALPHA:
        res = isAlpha(ch);
        break;
      case RI_LOWER:
        res = isLower(ch);
        break;
      case RI_UPPER:
        res = isUpper(ch);
        break;
      case RI_WHITE:
        res = isWhite(ch);
        break;
    }

    return (res == (test > 0));
  }

  public static boolean isHex(char ch) {
    return Character.digit(ch, 16) != -1;
  }

  public static boolean isOctal(char ch) {
    return Character.digit(ch, 8) != -1;
  }

  public static boolean isWord(char ch) {
    return Character.isLetterOrDigit(ch) || ch == '_';
  }

  public static boolean isHead(char ch) {
    return Character.isLetter(ch) || ch == '_';
  }

  public static boolean isAlpha(char ch) {
    return Character.isLetter(ch);
  }

  public static boolean isLower(char ch) {
    return Character.isLowerCase(ch);
  }

  public static boolean isUpper(char ch) {
    return Character.isUpperCase(ch);
  }

  public static boolean isWhite(char ch) {
    return Character.isWhitespace(ch);
  }

  public static boolean isGraph(char ch) {
    return (ch >= '!' && ch <= '~');
  }

  public static boolean isPrint(char ch) {
    return ((ch >= ' ' && ch <= '~') || ch > 0xff);
  }

  public static boolean isPunct(char ch) {
    return ((ch >= '!' && ch <= '/') ||
            (ch >= ':' && ch <= '@') ||
            (ch >= '[' && ch <= '`') ||
            (ch >= '{' && ch <= '~'));
  }

  public static boolean isFile(char ch) {
    return (isWord(ch) || "/.-+,#$%~=".indexOf(ch) != -1);
  }
}

File: vim-engine/src/main/kotlin/com/maddyhome/idea/vim/regexp/CharacterClasses.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.regexp

import org.jetbrains.annotations.NonNls

@Deprecated("Remove once old regex engine is removed")
object CharacterClasses {
  val CLASS_NAMES: @NonNls Array<String> = arrayOf(
    "alnum:]",
    "alpha:]",
    "blank:]",
    "cntrl:]",
    "digit:]",
    "graph:]",
    "lower:]",
    "print:]",
    "punct:]",
    "space:]",
    "upper:]",
    "xdigit:]",
    "tab:]",
    "return:]",
    "backspace:]",
    "escape:]",
  )
  const val RI_DIGIT: Int = 0x01
  const val RI_HEX: Int = 0x02
  const val RI_OCTAL: Int = 0x04
  const val RI_WORD: Int = 0x08
  const val RI_HEAD: Int = 0x10
  const val RI_ALPHA: Int = 0x20
  const val RI_LOWER: Int = 0x40
  const val RI_UPPER: Int = 0x80
  const val RI_WHITE: Int = 0x100
  const val CLASS_ALNUM: Int = 0
  const val CLASS_ALPHA: Int = 1
  const val CLASS_BLANK: Int = 2
  const val CLASS_CNTRL: Int = 3
  const val CLASS_DIGIT: Int = 4
  const val CLASS_GRAPH: Int = 5
  const val CLASS_LOWER: Int = 6
  const val CLASS_PRINT: Int = 7
  const val CLASS_PUNCT: Int = 8
  const val CLASS_SPACE: Int = 9
  const val CLASS_UPPER: Int = 10
  const val CLASS_XDIGIT: Int = 11
  const val CLASS_TAB: Int = 12
  const val CLASS_RETURN: Int = 13
  const val CLASS_BACKSPACE: Int = 14
  const val CLASS_ESCAPE: Int = 15
  const val CLASS_NONE: Int = 99
  fun isMask(ch: Char, mask: Int, test: Int): Boolean {
    var res = false
    when (mask) {
      RI_DIGIT -> res = Character.isDigit(ch)
      RI_HEX -> res = isHex(ch)
      RI_OCTAL -> res = isOctal(ch)
      RI_WORD -> res = isWord(ch)
      RI_HEAD -> res = isHead(ch)
      RI_ALPHA -> res = isAlpha(ch)
      RI_LOWER -> res = isLower(ch)
      RI_UPPER -> res = isUpper(ch)
      RI_WHITE -> res = isWhite(ch)
    }
    return res == test > 0
  }

  fun isHex(ch: Char): Boolean {
    return (ch.digitToIntOrNull(16) ?: -1) != -1
  }

  fun isOctal(ch: Char): Boolean {
    return (ch.digitToIntOrNull(8) ?: -1) != -1
  }

  fun isWord(ch: Char): Boolean {
    return Character.isLetterOrDigit(ch) || ch == '_'
  }

  fun isHead(ch: Char): Boolean {
    return Character.isLetter(ch) || ch == '_'
  }

  @JvmStatic
  fun isAlpha(ch: Char): Boolean {
    return Character.isLetter(ch)
  }

  fun isLower(ch: Char): Boolean {
    return Character.isLowerCase(ch)
  }

  fun isUpper(ch: Char): Boolean {
    return Character.isUpperCase(ch)
  }

  fun isWhite(ch: Char): Boolean {
    return Character.isWhitespace(ch)
  }

  fun isGraph(ch: Char): Boolean {
    return ch >= '!' && ch <= '~'
  }

  fun isPrint(ch: Char): Boolean {
    return ch >= ' ' && ch <= '~' || ch.code > 0xff
  }

  fun isPunct(ch: Char): Boolean {
    return ch in '!'..'/' ||
      ch in ':'..'@' ||
      ch in '['..'`' ||
      ch in '{'..'~'
  }

  fun isFile(ch: Char): Boolean {
    return isWord(ch) || "/.-+,#$%~=".indexOf(ch) != -1
  }
}
